core/src/main/java/org/apache/calcite/rel/core/Project.java

```diff
@@ -282,7 +282,12 @@ public RelWriter explainTerms(RelWriter pw) {
       if (!(exp.e instanceof RexInputRef)) {
         return null;
       }
```

Comment:
https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/rel/core/Project.java#L265
How do you think this line, is it safe ?

Reply:
I don't think it is safe to call `TargetMapping#getTarget(int)` mentioned in the comment.

Reply:
I think just changing the mapping type to `INVERSE_FUNCTION` and removing `Mappings.isIdentity(mapping)` check in L109 can also solve the issue. L275~L277 should be removed. Apparently, the mapping is not an inverse surjection, but an inverse function, according to the method description.

Reply:
What you describe is exactly what `getPartialMapping` do.